File: dashboard/src/modules/member/state/get-members-request.ts
import {
  useQueryStates,
  parseAsString,
  parseAsNumberLiteral,
  createSerializer,
  parseAsBoolean,
  parseAsIsoDateTime,
} from 'nuqs';
import { createSearchParamsCache } from 'nuqs/server';

import { SortOrder } from '@/types';

import { paginationParsers } from '@/core/data/state';

const parsers = {
  query: parseAsString,
  joinedFrom: parseAsIsoDateTime,
  joinedTo: parseAsIsoDateTime,
  isMississaugaResident: parseAsBoolean,
  sortFirstName: parseAsNumberLiteral(Object.values(SortOrder) as number[]),
  sortLastName: parseAsNumberLiteral(Object.values(SortOrder) as number[]),
  sortJoinedDate: parseAsNumberLiteral(Object.values(SortOrder) as number[]),
  ...paginationParsers,
};

export const useGetMembersUrlParams = () => {
  return useQueryStates(
    parsers,
    { shallow: false }
  );
};

export const getMembersUrlParams = createSearchParamsCache(parsers);

export const serializeGetMembersUrlParams = createSerializer(parsers);

File: api/MfaApi/src/Modules/Member/Contracts/GetMembersRequest.cs
using MfaApi.Core.Sort;
using MfaApi.Core.Pagination;

namespace MfaApi.Modules.Member;

public class GetMembersRequest: PaginationRequest {
    public string? Query { get; set; }
    public DateTime? JoinedFrom { get; set; }
    public DateTime? JoinedTo { get; set; }
    public bool? IsMississaugaResident { get; set; }
    public SortOrder? SortFirstName { get; set; }
    public SortOrder? SortLastName { get; set; }
    public SortOrder? SortJoinedDate { get; set; }
    public bool ShowArchived { get; set; } = false;
}